test_conformance/extensions/cl_khr_command_buffer/negative_command_buffer_copy.cpp

```diff
@@ -0,0 +1,659 @@
+struct CommandBufferCopyBufferQueueNotNull
+    : public CommandBufferCopyBaseTest<false>
+{
+    using CommandBufferCopyBaseTest::CommandBufferCopyBaseTest;
+
+    cl_int Run() override
+    {
+        cl_int error =
+            clCommandCopyBufferKHR(command_buffer, queue, in_mem, out_mem, 0, 0,
+                                   data_size, 0, nullptr, nullptr, nullptr);
+
+        test_failure_error_ret(error, CL_INVALID_COMMAND_QUEUE,
+                               "clCommandCopyBufferKHR should return "
+                               "CL_INVALID_COMMAND_QUEUE",
+                               TEST_FAIL);
+
+        error = clCommandCopyBufferRectKHR(
+            command_buffer, queue, in_mem, out_mem, origin, origin, region, 0,
+            0, 0, 0, 0, nullptr, nullptr, nullptr);
+
+        test_failure_error_ret(error, CL_INVALID_COMMAND_QUEUE,
+                               "clCommandCopyBufferRectKHR should return "
+                               "CL_INVALID_COMMAND_QUEUE",
+                               TEST_FAIL);
+
+        return CL_SUCCESS;
+    }
+};
+
+// CL_INVALID_COMMAND_QUEUE if command_queue is not NULL.
+struct CommandBufferCopyImageQueueNotNull
+    : public CommandBufferCopyBaseTest<true>
+{
+    using CommandBufferCopyBaseTest::CommandBufferCopyBaseTest;
+
+    cl_int Run() override
+    {
+        cl_int error = clCommandCopyImageToBufferKHR(
+            command_buffer, queue, image, buffer, origin, region, 0, 0, nullptr,
+            nullptr, nullptr);
+
+        test_failure_error_ret(error, CL_INVALID_COMMAND_QUEUE,
+                               "clCommandCopyImageToBufferKHR should return "
+                               "CL_INVALID_COMMAND_QUEUE",
+                               TEST_FAIL);
+
+        return CL_SUCCESS;
+    }
+};
+
+// CL_INVALID_CONTEXT if the context associated with command_queue,
+// command_buffer, src_buffer, and dst_buffer are not the same.
+struct CommandBufferCopyBufferDifferentContexts
+    : public CommandBufferCopyBaseTest<false>
+{
+    using CommandBufferCopyBaseTest::CommandBufferCopyBaseTest;
+
+    cl_int SetUp(int elements) override
+    {
+        cl_int error = CommandBufferCopyBaseTest::SetUp(elements);
+        test_error(error, "BasicCommandBufferTest::SetUp failed");
+
+        context1 = clCreateContext(0, 1, &device, nullptr, nullptr, &error);
+        test_error(error, "Failed to create context");
+
+        in_mem_ctx = clCreateBuffer(context1, CL_MEM_READ_ONLY, data_size,
+                                    nullptr, &error);
+        test_error(error, "clCreateBuffer failed");
+
+        out_mem_ctx = clCreateBuffer(context1, CL_MEM_WRITE_ONLY, data_size,
+                                     nullptr, &error);
+        test_error(error, "clCreateBuffer failed");
+
+        return CL_SUCCESS;
+    }
+
+    cl_int Run() override
+    {
+        cl_int error = clCommandCopyBufferKHR(
+            command_buffer, nullptr, in_mem_ctx, out_mem, 0, 0, data_size, 0,
```

Comment:
Should `out_mem` be getting initalized in `CommandBufferCopyBaseTest::SetUp()`? I'm setting it as a nullptr here resulting in a different error than expected

Reply:
@EwanC corrected. I thought I added initialization in setup..., my bad.